backend/app/db/crud/user.py

```diff
@@ -20,56 +23,104 @@ def get_user_by_email(db: Session, email: str) -> t.Optional[User]:
-    hashed_password = get_password_hash(user.password)
-    db_user = User(
-        first_name=user.first_name,
-        last_name=user.last_name,
-        email=user.email,
-        is_active=user.is_active,
-        is_superuser=user.is_superuser,
-        hashed_password=hashed_password,
-    )
+def create_user(db: Session, user: UserBase):
+    """Create a user."""
+    db_user = User(**user.dict())
     db.add(db_user)
     db.commit()
     db.refresh(db_user)
     return db_user
 
 
-def delete_user(db: Session, user_id: int):
+def deactivate_user(db: Session, user_id: int) -> User:
     user = get_user(db, user_id)
     if not user:
         raise HTTPException(status.HTTP_404_NOT_FOUND, detail="User not found")
-    db.delete(user)
+    user.is_active = False
+    user.hashed_password = None
+    db.add(user)
     db.commit()
     return user
 
 
-def edit_user(
-    db: Session, user_id: int, user: app.db.schemas.user.UserEdit
-) -> app.db.schemas.user.User:
+def edit_user(db: Session, user_id: int, user: UserBase) -> User:
     db_user = get_user(db, user_id)
     if not db_user:
         raise HTTPException(status.HTTP_404_NOT_FOUND, detail="User not found")
     update_data = user.dict(exclude_unset=True)
 
-    if user.password:
-        update_data["hashed_password"] = get_password_hash(user.password)
-        del update_data["password"]
-
     for key, value in update_data.items():
         setattr(db_user, key, value)
 
     db.add(db_user)
     db.commit()
     db.refresh(db_user)
     return db_user
+
+
+def activate_user(
+    db: Session,
+    user: User,
+    activation_token: PasswordResetToken,
+    password: str,
+) -> User:
+    """Activate a user.
+
+    Sets a password, and toggles activation flags on user and activation_token.
+    """
+
+    user.hashed_password = get_password_hash(password)
+    user.is_active = True
+    db.add(user)
+
+    activation_token.is_redeemed = True
+    db.add(activation_token)
+
+    db.commit()
+    db.refresh(user)
+
+    return user
+
+
+def get_password_reset_token(
+    db: Session,
+    token: str,
+) -> PasswordResetToken:
+    activation_token: Optional[PasswordResetToken] = (
+        db.query(PasswordResetToken)
+        .filter(PasswordResetToken.token == token)
+        .one_or_none()
+    )
+    if activation_token is None:
+        raise HTTPException(status_code=404, detail="Token not found")
+    if activation_token.expiry_ts < datetime.datetime.utcnow():
+        raise HTTPException(status_code=404, detail="Token expired")
+    if activation_token.is_redeemed:
+        raise HTTPException(status_code=404, detail="Token already redeemed")
+
+    return activation_token
+
+
+def create_password_reset_token(
+    db: Session,
+    user_id: int,
+) -> PasswordResetToken:
+    # TODO make configurable
+    future_date = get_password_reset_token_expiry_ts()
+
+    row = PasswordResetToken(
+        user_id=user_id, token=random_string(), expiry_ts=future_date
```

Comment:
Related to the above: this 12 character random string seems a little arbitrary, why is this not a generated JWT activation token from which we extract the appropriate data?

Reply:
Because it doesn't need to be? Just has to be unique, hard to guess.

Do you mean that using JWT means we can do away with the password_reset_token table? Store the expiry as data in the JWT? If so, then good idea.

Reply:
But it might also be a good idea to keep track of these resets, especially tokens that expired and haven't been redeemed, so we can reach out to these users, and perhaps check if emails are bouncing, etc.

Reply:
What I initially thought we could do was encode the user id & expiry in the token, then validate & extract them from it... I had thought we could do away with the table, but now that I think about it more that probably means we'd lose the ability to track when a token had been used (and invalidation on multiple requests etc) :disappointed: .